### PR TITLE
Cleanup search tree

### PR DESCRIPTION
**Pull request by dalg24**

### CODE REVIEW

File: packages/Search/test/tstLinearBVH.cpp [hunk 1]
```diff
@@ -60,13 +60,14 @@ TEUCHOS_UNIT_TEST_TEMPLATE_1_DECL( LinearBVH, tag_dispatching, NO )
...
 
     DataTransferKit::BVH<NO> bvh( boxes );
-    Kokkos::View<int *, DeviceType> results;
+    int results[255];
+    unsigned int n_results;
     DataTransferKit::Point p1 = {0., 0., 0.};
-    DataTransferKit::BVHQuery<NO>::query( bvh, details::nearest( p1, 1 ),
-                                          results );
+    details::TreeTraversal<NO>::query( bvh, details::nearest( p1, 1 ), results,
+                                       n_results );
```
Comment (aprokop) [inline packages/Search/test/tstLinearBVH.cpp]: How would this work? Your `results` live on the host, however, your `query` runs on the device. This results in invalid pointer, right?

**Comment by aprokop** [comment]
Related to #217.

**Comment by aprokop** [comment]
Fix #214.

**Comment by aprokop** [comment]
Rebased on top of master.

**Comment by aprokop** [comment]
@dalg24 Should the `query` in `tstLinearBVH` go to `src` somewhere, instead of sitting in `test`? It seems like a really convenient interface...

**Comment by aprokop** [comment]
Also, if you mark some of the query functions as deprecated, should not the code in `tstLinearBVH` be updated to remove them? Or do we just want to push this PR sooner rather than later?

**Comment by dalg24** [comment]
@aprokop I meant to turn `query` into a member function of `BVH`.  We should do this **before** we merge.

**Comment by aprokop** [comment]
So, the only thing remaining in this PR is to get rid of DEPRECATED `query` functions that @dalg24  is working on.

File: packages/Search/examples/bvh_driver/bvh_driver.cpp [hunk 2]
```diff
@@ -19,86 +19,7 @@
...
-                   _point_coords( i, 2 )},
-            _k[i], indices, n_indices, max_n_indices );
-    }
-
-  private:
-    PointsView _point_coords;
-    KView _k;
-    DataTransferKit::BVH<NO> _bvh;
-};
+namespace details = DataTransferKit::Details;
```
Comment (Rombur) [inline packages/Search/examples/bvh_driver/bvh_driver.cpp]: Do we want to keep the namespace or not?

File: packages/Search/src/DTK_LinearBVH_decl.hpp [hunk 3]
```diff
@@ -23,6 +23,12 @@
 
 namespace DataTransferKit
 {
+namespace Details
+{
```
Comment (Rombur) [inline packages/Search/src/DTK_LinearBVH_decl.hpp]: So this is part of the namespace details but not in the details subdirectory :-/

File: packages/Search/src/DTK_LinearBVH_decl.hpp [hunk 4]
```diff
@@ -32,19 +38,107 @@ struct BVH
...
-    Kokkos::View<Node *, DeviceType> leaf_nodes;
-    Kokkos::View<Node *, DeviceType> internal_nodes;
-    /**
-     *  Array of indices that sort the boxes used to construct the hierarchy.
-     *  The leaf nodes are ordered so we need these to identify objects that
-     * meet
-     *  a predicate.
+    template <typename Query>
+    void query( Kokkos::View<Query *, DeviceType> queries,
+                Kokkos::View<int *, DeviceType> &indices,
```
Comment (Rombur) [inline packages/Search/src/DTK_LinearBVH_decl.hpp]: We should comment why we need to pass by reference.

File: packages/Search/src/DTK_LinearBVH_decl.hpp [hunk 5]
```diff
@@ -32,19 +38,107 @@ struct BVH
...
+                Kokkos::View<int *, DeviceType> &offset ) const;
+
+  private:
+    friend struct Details::TreeTraversal<NO>;
+
+    Kokkos::View<Node *, DeviceType> _leaf_nodes;
+    Kokkos::View<Node *, DeviceType> _internal_nodes;
+    /*
```
Comment (Rombur) [inline packages/Search/src/DTK_LinearBVH_decl.hpp]: The comment won't be picked up by doxygen. Is this what we want?

File: packages/Search/src/DTK_LinearBVH_decl.hpp [hunk 6]
```diff
@@ -32,19 +38,107 @@ struct BVH
...
+
+    // Initialize view
+    // [ 0 0 0 .... 0 0 ]
+    //                ^
+    //                N
+    Kokkos::resize( offset, n_queries + 1 );
+    Kokkos::parallel_for(
+        "query(): initialize offset (set all entries to zero)",
+        Kokkos::RangePolicy<ExecutionSpace>( 0, n_queries + 1 ),
+        KOKKOS_LAMBDA( int i ) { offset[i]; } );
```
Comment (Rombur) [inline packages/Search/src/DTK_LinearBVH_decl.hpp]: Use `offeset[i]=0;` I am not sure the `parallel_for` does anything currently.

File: packages/Search/test/tstLinearBVH.cpp [hunk 7]
```diff
@@ -685,51 +585,79 @@ TEUCHOS_UNIT_TEST_TEMPLATE_1_DECL( LinearBVH, rtree, NO )
...
+    Kokkos::View<int *, DeviceType> offset_nearest( "offset_nearest" );
+    Kokkos::View<int *, DeviceType> indices_nearest( "indices_nearest" );
+    bvh.query( nearest_queries, indices_nearest, offset_nearest );
+
+    Kokkos::View<int *, DeviceType> offset_within( "offset_within" );
+    Kokkos::View<int *, DeviceType> indices_within( "indices_within" );
+    bvh.query( within_queries, indices_within, offset_within );
 
-    for ( int i = 0; i < n_points; ++i )
+    struct Dumpster
```
Comment (Rombur) [inline packages/Search/test/tstLinearBVH.cpp]: Maybe just use a `tuple` instead of creating a struct. Up to you.

File: packages/Search/examples/bvh_driver/bvh_driver.cpp (hunk 2 above)
Comment (aprokop) [inline packages/Search/examples/bvh_driver/bvh_driver.cpp]: I'd say yes.

File: packages/Search/src/DTK_LinearBVH_decl.hpp (hunk 3 above)
Comment (aprokop) [inline packages/Search/src/DTK_LinearBVH_decl.hpp]: We need to for friend declaration. No way around it.

File: packages/Search/src/DTK_LinearBVH_decl.hpp (hunk 5 above)
Comment (aprokop) [inline packages/Search/src/DTK_LinearBVH_decl.hpp]: Oh, that was my bad...

File: packages/Search/src/DTK_LinearBVH_decl.hpp (hunk 3 above)
Comment (Rombur) [inline packages/Search/src/DTK_LinearBVH_decl.hpp]: Then let's pick another name. If I see details, I will look into the directory `details`.

Comment (aprokop) [inline packages/Search/src/DTK_LinearBVH_decl.hpp]: I'm not sure I understand. This namespace only applies to a single declaration of `TreeTraversal`. The code for it lives in the `details` subdir, as expected.

Comment (Rombur) [inline packages/Search/src/DTK_LinearBVH_decl.hpp]: my bad.

File: packages/Search/src/DTK_LinearBVH_decl.hpp (hunk 5 above)
Comment (dalg24) [inline packages/Search/src/DTK_LinearBVH_decl.hpp]: What would be the proper way to do that?

Comment (aprokop) [inline packages/Search/src/DTK_LinearBVH_decl.hpp]: Revert this change, it used to be `/**`

**Comment by dalg24** [comment]
Ready for final review/merge.

**Comment by aprokop** [comment]
Looks good to me.
